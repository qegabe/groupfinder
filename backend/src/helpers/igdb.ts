import axios from "axios";

const IGDB_URL = "https://api.igdb.com/v4";

/**
 * Checks if the twitch token is valid
 * @returns {Promise<boolean>}
 */
async function validateToken(): Promise<boolean> {
  if (process.env.TWITCH_TOKEN !== undefined) {
    try {
      const res = await axios({
        method: "GET",
        url: `https://id.twitch.tv/oauth2/validate`,
        headers: {
          Authorization: `Bearer ${process.env.TWITCH_TOKEN}`,
        },
      });

      console.log(`Twitch validated with ${res.status}`);
      return true;
    } catch (error) {
      console.error(error);
      return false;
    }
  }

  return false;
}

/**
 * Send a request to IGDB
 * @param endpoint endpoint for request
 * @param data
 * @returns {Promise<any[]>}
 */
async function requestIGDB(endpoint: string, data: string): Promise<any[]> {
  console.log("Sending request to IGDB...");
  try {
    const resp = await axios({
      url: `${IGDB_URL}/${endpoint}`,
      method: "POST",
      headers: {
        Authorization: `Bearer ${process.env.TWITCH_TOKEN}`,
        "Client-ID": process.env.TWITCH_CLIENT_ID,
      },
      data,
    });
    return resp.data;
  } catch (error) {
    //console.error(error);
    console.error(`IGDB Request failed with: ${error.code}`);
  }
}

export { validateToken, requestIGDB };
