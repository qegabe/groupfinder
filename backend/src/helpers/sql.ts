import { BadRequestError } from "./expressError";
import { Filter } from "../types";

const jsToSql = {
  avatarUrl: "avatar_url",
  triviaScore: "trivia_score",
  isPrivate: "is_private",
  startTime: "start_time",
  endTime: "end_time",
  maxMembers: "max_members",
};

/**
 * Creates sql for updating a row
 * @param dataToUpdate
 * @returns {object} { setCols, values }
 */
function sqlForPartialUpdate(dataToUpdate: object) {
  const keys = Object.keys(dataToUpdate);
  if (keys.length === 0) throw new BadRequestError("No data");

  // {username: 'mike', avatarUrl: 'img.jpg'} => ['"username"=$1', '"avatar_url"=$2']
  const cols = keys.map(
    (colName, idx) =>
      `"${jsToSql[colName as keyof object] || colName}"=$${idx + 1}`
  );

  return {
    setCols: cols.join(", "),
    values: Object.values(dataToUpdate),
  };
}

/**
 * Creates sql for filtering a SELECT query
 * @param filter
 * @returns {object} { matchers, values }
 */
function sqlForFiltering(filter: Filter) {
  const keys = Object.keys(filter);

  const matchers = [];
  const values = [];
  for (let key of keys) {
    const i = values.length + 1;
    switch (key) {
      case "title":
        matchers.push(`title ILIKE $${i}`);
        values.push(`%${filter.title}%`);
        break;
      case "startTimeAfter":
        matchers.push(`start_time >= $${i}`);
        values.push(filter.startTimeAfter);
        break;
      case "startTimeBefore":
        matchers.push(`start_time <= $${i}`);
        values.push(filter.startTimeBefore);
        break;
      case "maxSize":
        const m = `id IN (SELECT id
                FROM groups
                LEFT JOIN groupsusers ON groupsusers.group_id = groups.id
                GROUP BY id
                HAVING COUNT(*) <= $${i}
              )`;
        matchers.push(m);
        values.push(filter.maxSize);
        break;
      default:
        break;
    }
  }

  return { matchers, values };
}

/**
 * Creates sql for the columns and values of an INSERT query
 * @param data
 * @returns {object} { colString, valString, values}
 */
function sqlForInserting(data: object) {
  const keys = Object.keys(data);

  const cols: string[] = [];
  const valueIdxs: string[] = [];

  for (let i = 0; i < keys.length; i++) {
    cols.push(jsToSql[keys[i] as keyof object] || keys[i]);
    valueIdxs.push(`$${i + 1}`);
  }

  return {
    colString: `(${cols.join(", ")})`,
    valString: `(${valueIdxs.join(", ")})`,
    values: Object.values(data),
  };
}

export { sqlForPartialUpdate, sqlForFiltering, sqlForInserting };
