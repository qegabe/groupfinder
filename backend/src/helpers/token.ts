import jwt from "jsonwebtoken";
import { SECRET_KEY } from "../config";
import { IUser } from "../types";

function createToken(user: IUser) {
  const payload = {
    username: user.username,
  };

  return jwt.sign(payload, SECRET_KEY);
}

export default createToken;
