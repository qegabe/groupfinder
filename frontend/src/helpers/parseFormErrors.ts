function simplifyError(error: string) {
  if (error.includes("User already exists with username")) {
    return { username: { error: true, helperText: "Username already taken" } };
  }
  return { message: error };
}

function parseFormErrors(error: string) {
  let errArray;
  try {
    errArray = JSON.parse(error);
  } catch {
    return simplifyError(error);
  }

  const errors: any = {};
  if (Array.isArray(errArray)) {
    for (let e of errArray) {
      const prop = e.match(/^instance\.(\w+)/)[1];
      if (prop) {
        if (e.includes("does not meet minimum length")) {
          errors[prop as keyof object] = {
            error: true,
            helperText: `Too Short, Must be atleast ${
              e[e.length - 1]
            } character(s) long`,
          };
        } else if (e.includes("is not of a type(s) string")) {
          errors[prop as keyof object] = {
            error: true,
            helperText: "Please Enter",
          };
        } else {
          errors[prop as keyof object] = {
            error: true,
          };
        }
      }
    }
  }

  return errors;
}

export default parseFormErrors;
