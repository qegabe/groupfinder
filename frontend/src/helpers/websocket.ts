const host =
  process.env.NODE_ENV === "production"
    ? window.location.host
    : "localhost:5000";

function startWebSocket(
  id: number,
  token: string,
  onMessageCallback: (data: any) => void
): WebSocket {
  const ws = new WebSocket(`ws://${host}/chat/${id}`);

  ws.onopen = () => {
    ws.send(JSON.stringify({ type: "join", token }));
    //console.log("opened ws connection");
  };

  ws.onclose = (e) => {
    console.log("close ws connection:", e.code, e.reason);
  };

  ws.onmessage = (e) => {
    onMessageCallback(e.data);
  };

  return ws;
}

export { startWebSocket };
