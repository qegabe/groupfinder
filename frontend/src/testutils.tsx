import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import { RootState, setupStore } from "./store";
import { Provider } from "react-redux";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import { LocalizationProvider } from "@mui/x-date-pickers";
import { AdapterDayjs } from "@mui/x-date-pickers/AdapterDayjs";
import { PreloadedState } from "@reduxjs/toolkit";

interface CustomRenderOptions {
  currentRoute?: string;
  routePath?: string;
  preloadedState?: PreloadedState<RootState>;
}

function customRender(
  ui: ReactElement,
  cOptions: CustomRenderOptions = {},
  options?: Omit<RenderOptions, "wrapper">
) {
  const store = setupStore(cOptions.preloadedState);
  function AllTheProviders({ children }: { children: React.ReactNode }) {
    return (
      <Provider store={store}>
        <MemoryRouter
          initialEntries={
            cOptions.currentRoute ? [cOptions.currentRoute] : undefined
          }>
          <LocalizationProvider dateAdapter={AdapterDayjs}>
            {cOptions.routePath ? (
              <Routes>
                <Route path={cOptions.routePath} element={children} />
              </Routes>
            ) : (
              children
            )}
          </LocalizationProvider>
        </MemoryRouter>
      </Provider>
    );
  }
  return render(ui, { wrapper: AllTheProviders, ...options });
}

export * from "@testing-library/react";
export { customRender as render };
