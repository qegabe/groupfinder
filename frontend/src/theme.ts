import { ThemeOptions, createTheme } from "@mui/material/styles";

const themeOptions: ThemeOptions = {
  palette: {
    mode: "dark",
    primary: {
      main: "#3fb56a",
    },
    secondary: {
      main: "#a5d6a7",
    },
    success: {
      main: "#64dd17",
    },
  },
};

export const theme = createTheme(themeOptions);
