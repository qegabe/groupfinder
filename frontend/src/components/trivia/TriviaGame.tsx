import React, { useRef, useMemo, useEffect, useState } from "react";
import { Box, Button, Grid, Typography } from "@mui/material";
import { startWebSocket } from "../../helpers/websocket";
import { useNavigate, useParams } from "react-router-dom";
import { useAppSelector } from "../../hooks/redux";
import TriviaQuestion from "./TriviaQuestion";
import GroupFinderApi from "../../api";
import TriviaScores from "./TriviaScores";
import TriviaGameOver from "./TriviaGameOver";
import TriviaRoundTransition from "./TriviaRoundTransition";

interface GameState {
  question: Question | undefined;
  correctAnswer: string | undefined;
  selectedAnswer: number | undefined;
  started: boolean;
  gameOver: boolean;
  round: number;
  showRound: boolean;
  scores: { [username: string]: number };
}

const INITIAL_STATE: GameState = {
  question: undefined,
  correctAnswer: undefined,
  selectedAnswer: undefined,
  started: false,
  gameOver: false,
  round: 0,
  showRound: false,
  scores: {},
};

function TriviaGame() {
  const { id } = useParams();
  const { token } = useAppSelector((s) => s.auth);
  const [groupUsers, setGroupUsers] = useState<Group["members"]>({});
  const [gameState, setGameState] = useState<GameState>(INITIAL_STATE);
  const navigate = useNavigate();

  const ws = useRef<WebSocket>();

  const messageRecieved = useMemo(
    () => (data: any) => {
      const msg = JSON.parse(data);

      switch (msg.type) {
        case "auth":
          if (!msg.result) {
            ws.current?.close();
            ws.current = undefined;
            navigate("/");
          }
          break;
        case "start":
          setGameState((gs) => ({
            ...gs,
            started: true,
          }));
          break;
        case "gameState":
          setGameState((gs) => ({
            ...gs,
            question: msg.question,
            correctAnswer: msg.correctAnswer,
            started: msg.started,
            scores: msg.scores,
            round: msg.round,
          }));
          break;
        case "question":
          setGameState((gs) => ({
            ...gs,
            question: msg.question,
            selectedAnswer: undefined,
            correctAnswer: undefined,
            showRound: false,
          }));
          break;
        case "result":
          setGameState((gs) => ({
            ...gs,
            answered: false,
            correctAnswer: msg.correctAnswer,
            scores: msg.scores,
          }));
          break;
        case "nextRound":
          setGameState((gs) => ({
            ...gs,
            round: msg.round,
            showRound: true,
          }));
          break;
        case "final":
          setGameState((gs) => ({
            ...gs,
            answered: false,
            gameOver: true,
            scores: msg.scores,
          }));
          break;
        case "restart":
          setGameState(INITIAL_STATE);
          break;
        default:
          break;
      }
    },
    [navigate]
  );

  //Connect to websocket
  useEffect(() => {
    if (!ws.current) {
      ws.current = startWebSocket(
        "trivia",
        +(id as string),
        token || "",
        messageRecieved
      );
    }
    return () => {
      if (ws.current?.readyState === ws.current?.OPEN) {
        ws.current?.close();
        ws.current = undefined;
      }
    };
  }, [token, id, messageRecieved]);

  //get user data
  useEffect(() => {
    async function getUsers() {
      const group = await GroupFinderApi.getGroup(+(id as string));
      setGroupUsers(group.members);
    }
    getUsers();
  }, [id]);

  function sendAnswer(idx: number) {
    ws.current?.send(
      JSON.stringify({
        type: "answer",
        answer: gameState.question?.answers[idx],
      })
    );
    setGameState((gs) => ({
      ...gs,
      selectedAnswer: idx,
    }));
  }

  function startGame() {
    ws.current?.send(JSON.stringify({ type: "start" }));
  }

  function nextQuestion() {
    ws.current?.send(JSON.stringify({ type: "next" }));
  }

  function restartGame() {
    ws.current?.send(JSON.stringify({ type: "restart" }));
  }

  const restartButton = (
    <Button variant="contained" color="error" onClick={restartGame}>
      Restart
    </Button>
  );

  if (gameState.gameOver)
    return (
      <TriviaGameOver
        users={groupUsers}
        scores={gameState.scores}
        restartButton={restartButton}
      />
    );

  let side = null;
  if (gameState.started && !gameState.showRound) {
    side = (
      <TriviaQuestion
        selectedAnswer={gameState.selectedAnswer}
        correctAnswer={gameState.correctAnswer}
        question={gameState.question}
        sendAnswer={sendAnswer}
        nextQuestion={nextQuestion}
      />
    );
  } else if (gameState.showRound) {
    side = <TriviaRoundTransition round={gameState.round} />;
  } else {
    side = (
      <Box
        sx={{
          display: "flex",
          flexDirection: "column",
        }}>
        <Typography variant="h1" sx={{ my: 4 }}>
          Trivia
        </Typography>
        <Button size="large" variant="contained" onClick={startGame}>
          Start Game
        </Button>
      </Box>
    );
  }

  return (
    <Box>
      <Grid container height={800}>
        <Grid
          sx={{ boxShadow: 1, p: 2, mt: 2 }}
          item
          container
          direction="column"
          justifyContent="flex-start"
          xs={3}>
          <Grid item xs={11}>
            <Typography variant="h4" display="flex" justifyContent="center">
              Scores
            </Typography>
            <Box>
              <TriviaScores users={groupUsers} scores={gameState.scores} />
            </Box>
          </Grid>
          <Grid item xs={1}>
            <Box
              sx={{
                display: "flex",
                justifyContent: "center",
              }}>
              {restartButton}
            </Box>
          </Grid>
        </Grid>
        <Grid
          item
          container
          xs={9}
          justifyContent="center"
          alignContent={gameState.started ? "baseline" : "center"}>
          {side}
        </Grid>
      </Grid>
    </Box>
  );
}

export default TriviaGame;
