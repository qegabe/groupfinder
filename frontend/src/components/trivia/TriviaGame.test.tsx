import React from "react";
import { render, waitFor, screen, fireEvent, within } from "../../testutils";
import "@testing-library/jest-dom";
import GroupFinderApi from "../../api";
import { WS } from "jest-websocket-mock";
import * as WSHelper from "../../helpers/websocket";
import dayjs from "dayjs";
import TriviaGame from "./TriviaGame";
import { theme } from "../../theme";

const mockGetGroup = jest
  .spyOn(GroupFinderApi, "getGroup")
  .mockImplementation(async (id) => {
    return {
      id: 1,
      title: "Test Group",
      description: "test",
      startTime: dayjs("2023-06-23T16:00:00.000Z"),
      endTime: dayjs("2023-06-23T17:00:00.000Z"),
      address: "123 Test Street",
      city: "Test City",
      cityId: 1,
      isPrivate: false,
      maxMembers: 10,
      members: {
        TestUser: {
          avatarUrl: null,
          isOwner: true,
        },
        TestUser3: {
          avatarUrl: null,
          isOwner: false,
        },
      },
      games: [
        { id: 1, title: "Test Game 1", coverUrl: "" },
        { id: -1, title: "Trivia", coverUrl: "" },
      ],
    };
  });

let WSServer: WS;

const mockStartWebSocket = jest
  .spyOn(WSHelper, "startWebSocket")
  .mockImplementation((path, id, token, onMessageCallback) => {
    const client = new WebSocket("ws://localhost:1234");
    client.onmessage = (e) => {
      onMessageCallback(e.data);
    };
    return client;
  });

beforeEach(() => {
  WSServer = new WS("ws://localhost:1234");
});

afterEach(() => {
  jest.clearAllMocks();
  WS.clean();
});

it("renders", async () => {
  render(<TriviaGame />, {
    currentRoute: "/games/trivia/1",
    routePath: "/games/trivia/:id",
  });
  await waitFor(() => {
    expect(screen.getAllByText("TestUser:")[0]).toBeInTheDocument();
  });
  expect(mockGetGroup).toHaveBeenCalledWith(1);
});

it("matches snapshot", async () => {
  const { asFragment } = render(<TriviaGame />, {
    currentRoute: "/games/trivia/1",
    routePath: "/games/trivia/:id",
  });
  await waitFor(() => {
    expect(screen.getAllByText("TestUser:")[0]).toBeInTheDocument();
  });
  expect(asFragment()).toMatchSnapshot();
});

it("can start game", async () => {
  render(<TriviaGame />, {
    currentRoute: "/games/trivia/1",
    routePath: "/games/trivia/:id",
  });
  await WSServer.connected;
  const startButton = screen.getByText("Start Game");
  fireEvent.click(startButton);
  await waitFor(() => {
    expect(WSServer).toReceiveMessage(JSON.stringify({ type: "start" }));
  });
});

it("can restart game", async () => {
  render(<TriviaGame />, {
    currentRoute: "/games/trivia/1",
    routePath: "/games/trivia/:id",
  });
  await WSServer.connected;
  const restartButton = screen.getAllByText("Restart")[0];
  fireEvent.click(restartButton);
  await waitFor(() => {
    expect(WSServer).toReceiveMessage(JSON.stringify({ type: "restart" }));
  });
});

it("displays round transition", async () => {
  render(<TriviaGame />, {
    currentRoute: "/games/trivia/1",
    routePath: "/games/trivia/:id",
  });
  await WSServer.connected;
  WSServer.send(JSON.stringify({ type: "start" }));
  WSServer.send(JSON.stringify({ type: "nextRound", round: 1 }));
  await waitFor(() => {
    expect(screen.getByText("Round 1")).toBeInTheDocument();
  });
});

it("displays questions and sends answers", async () => {
  render(<TriviaGame />, {
    currentRoute: "/games/trivia/1",
    routePath: "/games/trivia/:id",
  });
  await WSServer.connected;
  WSServer.send(JSON.stringify({ type: "start" }));
  WSServer.send(JSON.stringify({ type: "nextRound", round: 1 }));
  WSServer.send(
    JSON.stringify({
      type: "question",
      question: {
        question: "Test Question",
        answers: ["Answer 1", "Answer 2", "Answer 3", "Answer 4"],
        category: "Test Category",
      },
    })
  );
  await waitFor(() => {
    expect(screen.getByText("Test Question")).toBeInTheDocument();
  });
  const answerButton = screen.getByText("Answer 1");
  fireEvent.click(answerButton);
  await waitFor(() => {
    expect(WSServer).toReceiveMessage(
      JSON.stringify({ type: "answer", answer: "Answer 1" })
    );
  });
});

it("displays question results and request next question", async () => {
  render(<TriviaGame />, {
    currentRoute: "/games/trivia/1",
    routePath: "/games/trivia/:id",
  });
  await WSServer.connected;
  WSServer.send(JSON.stringify({ type: "start" }));
  WSServer.send(JSON.stringify({ type: "nextRound", round: 1 }));
  WSServer.send(
    JSON.stringify({
      type: "question",
      question: {
        question: "Test Question",
        answers: ["Answer 1", "Answer 2", "Answer 3", "Answer 4"],
        category: "Test Category",
      },
    })
  );
  await waitFor(() => {
    expect(screen.getByText("Test Question")).toBeInTheDocument();
  });
  WSServer.send(
    JSON.stringify({
      type: "result",
      correctAnswer: "Answer 1",
      scores: { TestUser: 100 },
    })
  );
  await waitFor(() => {
    expect(screen.getByText("Next Question")).toBeInTheDocument();
  });
  const correctAnswerButton = screen.getByText("Answer 1");
  expect(correctAnswerButton.className).toContain("MuiButton-containedSuccess");
  const { getByText: getScoreText } = within(
    screen.getAllByTestId("TestUser-score")[0]
  );
  expect(getScoreText("100")).toBeInTheDocument();
  fireEvent.click(screen.getByText("Next Question"));
  await waitFor(() => {
    expect(WSServer).toReceiveMessage(JSON.stringify({ type: "next" }));
  });
});

it("displays final scores", async () => {
  render(<TriviaGame />, {
    currentRoute: "/games/trivia/1",
    routePath: "/games/trivia/:id",
  });
  await WSServer.connected;
  WSServer.send(JSON.stringify({ type: "final", scores: { TestUser: 100 } }));
  await waitFor(() => {
    expect(screen.getByText("Final Scores")).toBeInTheDocument();
  });
  const { getByText: getScoreText } = within(
    screen.getByTestId("TestUser-score")
  );
  expect(getScoreText("100")).toBeInTheDocument();
});

it("get gamestate from server", async () => {
  render(<TriviaGame />, {
    currentRoute: "/games/trivia/1",
    routePath: "/games/trivia/:id",
  });
  await WSServer.connected;
  WSServer.send(
    JSON.stringify({
      type: "gameState",
      question: {
        question: "Test Question",
        answers: ["Answer 1", "Answer 2", "Answer 3", "Answer 4"],
        category: "Test Category",
      },
      correctAnswer: undefined,
      scores: { TestUser: 100 },
      round: 1,
      started: true,
    })
  );
  await waitFor(() => {
    expect(screen.getByText("Test Question")).toBeInTheDocument();
  });
  expect(screen.queryByText("Next Question")).not.toBeInTheDocument();
  const { getByText: getScoreText } = within(
    screen.getAllByTestId("TestUser-score")[0]
  );
  expect(getScoreText("100")).toBeInTheDocument();
});
