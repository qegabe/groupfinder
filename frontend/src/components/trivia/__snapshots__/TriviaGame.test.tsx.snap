// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-[object Object] css-11x07ja-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-md-3 css-c5kyim-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-11 css-ykovpn-MuiGrid-root"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 css-15bxn6s-MuiTypography-root"
          >
            Scores
          </h4>
          <div
            class="MuiBox-root css-0"
          >
            <div
              class="MuiStack-root css-1okj3ks-MuiStack-root"
            >
              <div
                class="MuiBox-root css-70qvj9"
              >
                <div
                  class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-cjda7z-MuiAvatar-root"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
                    data-testid="PersonIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                    />
                  </svg>
                </div>
                <span
                  class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
                >
                  TestUser:
                </span>
                <span
                  class="MuiTypography-root MuiTypography-body1 css-yyvcc2-MuiTypography-root"
                  data-testid="TestUser-score"
                >
                  0
                </span>
              </div>
              <div
                class="MuiBox-root css-70qvj9"
              >
                <div
                  class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-cjda7z-MuiAvatar-root"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
                    data-testid="PersonIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                    />
                  </svg>
                </div>
                <span
                  class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
                >
                  TestUser3:
                </span>
                <span
                  class="MuiTypography-root MuiTypography-body1 css-yyvcc2-MuiTypography-root"
                  data-testid="TestUser3-score"
                >
                  0
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-u78q51-MuiGrid-root"
        >
          <div
            class="MuiBox-root css-1l4w6pd"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium css-1sszw6i-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Restart
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-9 css-gzxg63-MuiGrid-root"
      >
        <div
          class="MuiBox-root css-j7qwjs"
        >
          <h1
            class="MuiTypography-root MuiTypography-h1 css-16s2ett-MuiTypography-root"
          >
            Trivia
          </h1>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge css-jh47zj-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Start Game
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root css-1deau0x"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Scores
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <div
        class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
        style="min-height: 0px;"
      >
        <div
          class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
        >
          <div
            class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
          >
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-nj1oux-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-11 css-ykovpn-MuiGrid-root"
              >
                <h4
                  class="MuiTypography-root MuiTypography-h4 css-15bxn6s-MuiTypography-root"
                >
                  Scores
                </h4>
                <div
                  class="MuiBox-root css-0"
                >
                  <div
                    class="MuiStack-root css-1okj3ks-MuiStack-root"
                  >
                    <div
                      class="MuiBox-root css-70qvj9"
                    >
                      <div
                        class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-cjda7z-MuiAvatar-root"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
                          data-testid="PersonIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                          />
                        </svg>
                      </div>
                      <span
                        class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
                      >
                        TestUser:
                      </span>
                      <span
                        class="MuiTypography-root MuiTypography-body1 css-yyvcc2-MuiTypography-root"
                        data-testid="TestUser-score"
                      >
                        0
                      </span>
                    </div>
                    <div
                      class="MuiBox-root css-70qvj9"
                    >
                      <div
                        class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-cjda7z-MuiAvatar-root"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
                          data-testid="PersonIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                          />
                        </svg>
                      </div>
                      <span
                        class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
                      >
                        TestUser3:
                      </span>
                      <span
                        class="MuiTypography-root MuiTypography-body1 css-yyvcc2-MuiTypography-root"
                        data-testid="TestUser3-score"
                      >
                        0
                      </span>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-u78q51-MuiGrid-root"
              >
                <div
                  class="MuiBox-root css-1l4w6pd"
                >
                  <button
                    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium css-1sszw6i-MuiButtonBase-root-MuiButton-root"
                    tabindex="0"
                    type="button"
                  >
                    Restart
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
